src/codegen/codegen_neuron_cpp_visitor.cpp

```diff
@@ -405,6 +410,73 @@ void CodegenNeuronCppVisitor::print_mechanism_global_var_structure(bool print_in
                       info.pointer_variables.size() > 0
                           ? static_cast<int>(info.pointer_variables.size())
                           : -1);
+
+    // Start printing the CNRN-style global variables.
+    auto float_type = default_float_data_type();
+    printer->add_newline(2);
+    printer->add_line("/** all global variables */");
+    printer->fmt_push_block("struct {}", global_struct());
+
+    if (!info.ions.empty()) {
+        // TODO implement these when needed.
+    }
+
+    if (info.point_process) {
+        throw std::runtime_error("Not implemented, global point process.");
+    }
+
+    if (!info.vectorize && !info.top_local_variables.empty()) {
+        throw std::runtime_error("Not implemented, global vectorize something.");
+    }
+
+    if (!info.thread_variables.empty()) {
+        throw std::runtime_error("Not implemented, global thread variables.");
+    }
+
+    if (info.table_count > 0) {
+        throw std::runtime_error("Not implemented, global table count.");
+    }
+
+    for (const auto& var: info.state_vars) {
```

Comment:
I'm wondering whether those parts of this function (or others) that are the same as `CodegenCoreneuronCppVisitor` should be moved into some function that be called by both visitors instead of duplicating the code

Reply:
Midterm, yes. The hope is that as we add more and more, the two implementations will automatically become more similar again. Once we fill out (and test) all of the

        throw std::runtime_error("Not implemented, global thread variables.");

we'd be in a better place to figure out the differences between NRN and CNRN (I think it's `inst->` vs `inst.` or something similarly minor) and solve it neatly.

Reply:
I hope the same as well. We can keep it as it is for now and hopefully we can get rid of more stuff in the end